control_panel_api/tests/test_aws_s3_access_policy.py

```diff
@@ -0,0 +1,210 @@
+
+    def assert_cannot_write(self, document, resource):
+        for action in self.WRITE_ACTIONS:
+            self.assert_not_allowed(document, action, f'{resource}/*')
+
+
+class S3AccessPolicyTestCase(S3AccessTestCase):
+
+    def test_blank_policy(self):
+        policy = S3AccessPolicy()
+        document = policy.document
+
+        self.assert_can_access_console(document)
+        self.assert_has_not_sid(document, 'list')
+        self.assert_has_not_sid(document, 'readonly')
+        self.assert_has_not_sid(document, 'readwrite')
+
+    def test_granting_access(self):
+        arns = [
+            ('arn:aws:s3:::readonly-1', False),
+            ('arn:aws:s3:::readonly-2', False),
+            ('arn:aws:s3:::readwrite-1', True),
+            ('arn:aws:s3:::readwrite-2', True),
+        ]
+
+        policy = S3AccessPolicy()
+        for (arn, readwrite) in arns:
+            policy.grant_access(arn, readwrite=readwrite)
+
+        document = policy.document
+        self.assert_can_access_console(document)
+
+        for (arn, readwrite) in arns:
+            self.assert_can_list(document, arn)
+            self.assert_can_read(document, arn)
+            if readwrite:
+                self.assert_can_write(document, arn)
+            else:
+                self.assert_cannot_write(document, arn)
+
+    def test_granting_access_again_with_different_level(self):
+        ARN = 'arn:aws:s3:::test-bucket-1'
+
+        policy = S3AccessPolicy()
+        policy.grant_access(ARN, readwrite=True)
+
+        document = policy.document
+        self.assert_can_access_console(document)
+        self.assert_can_list(document, ARN)
+        self.assert_can_read(document, ARN)
+        self.assert_can_write(document, ARN)
+
+        # Granting readonly access
+        policy.grant_access(ARN, readwrite=False)
+
+        document = policy.document
+        self.assert_can_access_console(document)
+        self.assert_can_list(document, ARN)
+        self.assert_can_read(document, ARN)
+        self.assert_cannot_write(document, ARN)
+
+    def test_revoking_access(self):
+        readonly_arn = 'arn:aws:s3:::readonly-1'
+        readwrite_arn = 'arn:aws:s3:::readwrite-1'
+
+        policy = S3AccessPolicy()
+        policy.grant_access(readonly_arn, readwrite=False)
+        policy.grant_access(readwrite_arn, readwrite=True)
+
+        document = policy.document
+        self.assert_can_access_console(document)
+        for arn in [readwrite_arn, readwrite_arn]:
+            self.assert_can_list(document, arn)
+            self.assert_can_read(document, arn)
+        self.assert_can_write(document, readwrite_arn)
+
+        policy.revoke_access(readwrite_arn)
+
+        document = policy.document
+        # access to `readwrite_arn` is now revoked
+        self.assert_cannot_list(document, readwrite_arn)
+        self.assert_cannot_read(document, readwrite_arn)
+        self.assert_cannot_write(document, readwrite_arn)
```

Comment:
The comment suggests the group of asserts could be more intent-revealing. Something like `self.assert_access_revoked(document, readwrite_arn)` or `self.assert_readonly_access(document, readwrite_arn)`?

Reply:
I could add a `assert_has_no_access()` assertion or something like that, but as this would be only used in this test I'll pass for this specific one.

I added `assert_has_readwrite_access()` and `assert_has_readonly_access()` which are used in several places and make the tests a joy to read :)